File: server/tasks_discarded/epds.py
#!/usr/bin/env python3
# epds.py

"""
    Copyright (C) 2012-2016 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

from cc_modules.cc_db import repeat_fieldspec
from cc_modules.cc_html import get_yes_no
from cc_modules.cc_string import WSTRING
from cc_modules.cc_task import get_from_dict, Task


# =============================================================================
# EPDS
# =============================================================================

class Epds(Task):
    NQUESTIONS = 10

    tablename = "epds"
    shortname = "EPDS"
    longname = "Edinburgh Postnatal Depression Scale"
    fieldspecs = repeat_fieldspec("q", 1, NQUESTIONS)

    TASK_FIELDS = [x["name"] for x in fieldspecs]

    def get_trackers(self):
        return [
            {
                "value": self.total_score(),
                "plot_label": "EPDS total score (rating depressive symptoms)",
                "axis_label": "Total score (out of 30)",
                "axis_min": -0.5,
                "axis_max": 30.5,
                "horizontal_lines": [
                    12.5,
                    9.5,
                ],
                "horizontal_labels": [
                    (13, "likely depression"),
                    (10, "possible depression"),
                ]
            }
        ]

    def get_summaries(self):
        return [
            self.is_complete_summary_field(),
            dict(name="total", cctype="INT",
                 value=self.total_score(), comment="Total score"),
        ]

    def is_complete(self):
        return self.are_all_fields_complete(self.TASK_FIELDS)

    def total_score(self):
        return self.sum_fields(self.TASK_FIELDS)

    def get_task_html(self):
        score = self.total_score()
        above_cutoff_1 = score >= 10
        above_cutoff_2 = score >= 13
        ANSWER_DICTS = []
        for q in range(1, self.NQUESTIONS + 1):
            d = {None: "?"}
            for option in range(0, 4):
                d[option] = (
                    str(option) + " — " +
                    WSTRING("epds_q" + str(q) + "_option" + str(option)))
            ANSWER_DICTS.append(d)
        h = """
            <div class="summary">
                <table class="summary">
                    {}
                    <tr><td>{}</td><td><b>{}</b> / 30</td></tr>
                    <tr><td>{} <sup>[1]</sup></td><td><b>{}</b></td></tr>
                    <tr><td>{} <sup>[2]</sup></td><td><b>{}</b></td></tr>
                </table>
            </div>
            <div class="explanation">
                Ratings are over the last week.
                <b>{}</b>
            </div>
            <table class="taskdetail">
                <tr>
                    <th width="50%">Question</th>
                    <th width="50%">Answer</th>
                </tr>
        """.format(
            self.get_is_complete_tr(),
            WSTRING("total_score"), score,
            WSTRING("epds_above_cutoff_1"), get_yes_no(above_cutoff_1),
            WSTRING("epds_above_cutoff_2"), get_yes_no(above_cutoff_2),
            WSTRING("epds_always_look_at_suicide")
        )
        for q in range(1, self.NQUESTIONS + 1):
            h += """<tr><td>{}</td><td><b>{}</b></td></tr>""".format(
                WSTRING("epds_q" + str(q) + "_question"),
                get_from_dict(ANSWER_DICTS[q - 1], getattr(self, "q" + str(q)))
            )
        h += """
            </table>
            <div class="footnotes">
                [1] &ge;10.
                [2] &ge;13.
                (Cox et al. 1987, PubMed ID 3651732.)
            </div>
        """
        return h

File: tablet/Resources/tasks_discarded/Epds.js
// Epds.js

/*
    Copyright (C) 2012-2015 Rudolf Cardinal (rudolf@pobox.com).
    Department of Psychiatry, University of Cambridge.
    Funded by the Wellcome Trust.

    This file is part of CamCOPS.

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

/*jslint node: true, newcap: true, nomen: true, plusplus: true */
"use strict";
/*global Titanium, L */

var DBCONSTANTS = require('common/DBCONSTANTS'),
    dbcommon = require('lib/dbcommon'),
    taskcommon = require('lib/taskcommon'),
    tablename = "epds",
    fieldlist = dbcommon.standardTaskFields(),
    nquestions = 10;

dbcommon.appendRepeatedFieldDef(fieldlist, "q", 1, nquestions, DBCONSTANTS.TYPE_INTEGER);

// CREATE THE TABLE

dbcommon.createTable(tablename, fieldlist);

// TASK

module.exports = function Epds(patient_id) {
    var self = this;
    taskcommon.createStandardTaskMembers(self, Epds, {
        tablename: tablename,
        fieldlist: fieldlist,
        patient_id: patient_id
    });

    // Scoring
    function getTotalScore() {
        return taskcommon.totalScore(self, "q", 1, nquestions);
    }

    // Standard task functions
    self.isComplete = function () {
        return taskcommon.isComplete(self, "q", 1, nquestions);
    };
    self.getSummary = function () {
        return (
            L('total_score') + " " + getTotalScore() + "/30, " +
            L('epds_suicidality') + " " + self.q10 + "/3" +
            self.isCompleteSuffix()
        );
    };
    self.getDetail = function () {
        var uifunc = require('lib/uifunc');
        return (
            taskcommon.valueDetail(self, "epds_q", "_s", " ", "q", 1, nquestions) +
            "\n" +
            self.getSummary() + "\n" +
            "\n" +
            L('epds_above_cutoff_1') + " " + uifunc.yesNo(getTotalScore() >= 10) + "\n" +
            L('epds_above_cutoff_2') + " " + uifunc.yesNo(getTotalScore() >= 13) + "\n" +
            L('epds_always_look_at_suicide')
        );
    };

    self.edit = function (readOnly) {
        var KeyValuePair = require('lib/KeyValuePair'),
            QuestionMCQ = require('questionnaire/QuestionMCQ'),
            Questionnaire = require('questionnaire/Questionnaire'),
            QuestionText = require('questionnaire/QuestionText'),
            pages,
            questionnaire;

        function makepage(n, reverse) {
            var options = [],
                i;
            if (reverse) {
                for (i = 3; i >= 0; --i) {
                    options.push(new KeyValuePair(L("epds_q" + n + "_option" + i), i));
                }
            } else {
                for (i = 0; i <= 3; ++i) {
                    options.push(new KeyValuePair(L("epds_q" + n + "_option" + i), i));
                }
            }
            return {
                title: L("epds_q" + n + "_title"),
                elements: [
                    { type: QuestionText, text: L("epds_question_common") },
                    { type: QuestionText, text: L("epds_q" + n + "_question"), bold: true },
                    { type: QuestionMCQ, field: "q" + n, options: options }
                ]
            };
        }
        pages = [];
        pages.push(makepage(1, false));
        pages.push(makepage(2, false));
        pages.push(makepage(3, true));
        pages.push(makepage(4, false));
        pages.push(makepage(5, true));
        pages.push(makepage(6, true));
        pages.push(makepage(7, true));
        pages.push(makepage(8, true));
        pages.push(makepage(9, true));
        pages.push(makepage(10, true));

        questionnaire = new Questionnaire({
            readOnly: readOnly,
            pages: pages,
            callbackThis: self,
            fnGetFieldValue: self.defaultGetFieldValueFn,
            fnSetField: self.defaultSetFieldFn,
            fnFinished: self.defaultFinishedFn
        });
        questionnaire.open();
    };

    return self;
};
